### PR TITLE
fix(e2b): atomic Resume with placeholder pausetime and min timeout floor

### PR DESCRIPTION
**Pull request by AiRanthem**
## Summary

The controller's auto-pause branch used to patch `Spec.Paused=true` without coordinating with the asymmetric `Resume()` in sandbox-manager (which used to flip `Paused=false` first and write the new `PauseTime` later, in a separate update). Between those two writes, the next reconcile observed `{Paused=false, PauseTime=stale}` and immediately re-paused the sandbox, breaking Resume.

This PR closes the race at the source — `Resume()` now writes `Paused=false` together with a placeholder timeout in a single Sandbox update — and tightens the controller-side patch to surface lost races cleanly. With the source-side fix in place, the previously-planned controller grace gate (`canFlipPausedTrue` + 30s Ready-stability window) is no longer needed and the `UpdatePolicyBaselineAware` workaround is removed.

### sandbox-manager Resume atomicity (`pkg/sandbox-manager/infra/{interface.go,sandboxcr/sandbox.go}`)

- `ResumeOptions` gains `Timeout *timeout.Options`. When set, `Resume()` writes `Spec.Paused=false` AND the placeholder `Spec.PauseTime` / `Spec.ShutdownTime` in the same `Update` (inside the existing `retryUpdate` mutator). The controller can no longer observe a stale deadline post-Resume.
- Mutator preserves first-writer-wins: only the caller that actually flips `Paused: true→false` writes the placeholder; losing concurrent callers short-circuit and never overwrite the winner's deadline.
- Never-timeout sandboxes pass `Timeout == nil` so Resume does not invent a deadline for them.

### Controller auto-pause patch (`pkg/controller/sandbox/sandbox_controller.go`)

- The auto-pause patch uses `client.MergeFromWithOptimisticLock`. A lost OCC race now surfaces as 409 → `Requeue: true` instead of silently overwriting concurrent state.
- `pauseTimeReached(pauseTime, now)` helper extracted; treats equal-time as due (`<=`).
- Branch only schedules `RequeueAfter` for a future PauseTime; if the time is reached, the patch is the only action — no partial writes.

### E2B handler timeout floor (`pkg/servers/e2b/{pause_resume.go,core.go,models/consts.go}`)

- New `applyResumeFloor`: bumps the requested timeout up to `--e2b-min-resume-timeout` (default 300s) when the request will trigger a Resume on a timed paused sandbox. Skipped for never-timeout sandboxes and for Running-state Connect.
- New `buildResumeOpts`: builds the atomic placeholder for `Resume()`; returns empty opts for never-timeout sandboxes so Resume stays a no-op write on those.
- Post-Resume timeout write goes through `updateConnectTimeout` with `UpdatePolicyExtendOnly`. Concurrent Connect/Resume callers therefore converge to the longest effective deadline; a shorter request cannot shrink a longer in-flight deadline.
- `ResumeSandbox` is now idempotent: an already-Running sandbox returns 204 instead of 409. `resumeSandboxErrorCode` maps NotFound → 404 and infra/cache conflict → 409 per the E2B `/resume` spec (no 400).

### Removed: `UpdatePolicyBaselineAware` (`pkg/utils/timeout/types.go`)

The Baseline-aware policy was introduced earlier to detect concurrent writers on a single timeout update; the atomic Resume placeholder + ExtendOnly post-write supersede it. The policy, the `Options.Baseline` field, and the associated tests are removed.

### CLI flag (`cmd/sandbox-manager/main.go`)

`--e2b-min-resume-timeout` (default 300s) with `validateE2BTimeoutFlags` (must be > 0 and <= `--e2b-max-timeout`).

### Docs

`pkg/servers/e2b/AGENTS.md` updated to match the new timeout semantics (atomic Resume placeholder, ExtendOnly post-Resume, no resume floor for running-state Connect).

## Tests

**Unit**
- `pkg/controller/sandbox/sandbox_controller_test.go`
  - `TestSandboxReconciler_AutoPauseBranch` (table-driven, 3 cases): OCC conflict -> Requeue with spec unchanged; successful patch persists `Paused=true`; future PauseTime -> `RequeueAfter` only.
  - `TestPauseTimeReached` (table-driven, 3 cases): past / equal / future.
- `pkg/sandbox-manager/infra/sandboxcr/pause_resume_test.go`
  - `TestSandbox_ResumeMutatorAtomicity`: placeholder + Paused=false are written in the same Update.
  - `TestSandbox_ResumeExtendOnlyConcurrent`: losing caller does not overwrite the winner's deadline.
  - `TestSandbox_ResumeFailurePathsLeaveRealTimeout`: failure paths do not clobber existing timeout state.
- `pkg/servers/e2b/pause_resume_test.go`
  - `TestResumeSandboxErrorCode` (table-driven): 404/409/500 mapping.
  - `TestConnectSandbox_ResumeFloorAndPlaceholder`, `TestResumeSandbox_ResumeFloorAndPlaceholder`: floor application + placeholder build.
- `cmd/sandbox-manager/main_test.go`: `TestValidateE2BTimeoutFlags` (table-driven): boundary cases for the new flag.

**E2E**
- `test/e2b/test_sandbox.py::test_auto_pause_resume_no_immediate_repause`:
  1. Create sandbox with `lifecycle={"on_timeout": "pause"}, timeout=30`.
  2. Wait for `Spec.Paused=true`.
  3. `connect_sandbox(sbx, timeout=600)`.
  4. Assert `state == RUNNING` for 15s after connect (no immediate re-pause).
  5. Assert `Spec.PauseTime ~ connect_start + 600s` (fresh deadline written by Resume, not stale leftover).

## Test plan

- [x] `go test ./pkg/controller/sandbox/... ./pkg/sandbox-manager/infra/sandboxcr/... ./pkg/servers/e2b/... ./pkg/utils/timeout/... ./cmd/sandbox-manager/...`
- [x] `go build ./...`
- [ ] E2E pytest against a Kind cluster (`make test-e2b`)
- [ ] Soak in staging: manual pause/resume and explicit `SetTimeout` paths must remain unaffected

## Out of scope

- Cross-replica route propagation through `syncRoute` gossip on the controller-driven path (separate from this fix).
- Refactoring `SaveTimeoutWithPolicy`'s retry loop to share with `retryUpdate` and switch to `APIReader`-on-conflict (raised in PR #358 review; revisit when the e2b layer adds more policies).

### CODE REVIEW

File: pkg/controller/sandbox/sandbox_controller.go [hunk 1]
```diff
@@ -210,13 +215,40 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (cr
 	}
 	if box.Spec.PauseTime != nil && !box.Spec.Paused {
 		if box.Spec.PauseTime.Before(&now) {
-			klog.InfoS("sandbox pause time reached, will be paused", "sandbox", klog.KObj(box))
-			modified := box.DeepCopy()
-			patch := client.MergeFrom(box)
-			modified.Spec.Paused = true
-			return ctrl.Result{}, r.Patch(ctx, modified, patch)
+			if !canFlipPausedTrue(box, now) {
```
Comment (furykerry) [inline pkg/controller/sandbox/sandbox_controller.go]: if sandbox manager patch pause=false and pauseTime in one patch, is canFlipPausedTrue still necessary?

File: pkg/controller/sandbox/sandbox_controller.go [hunk 2]
```diff
@@ -210,13 +215,40 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (cr
...
+				modified := box.DeepCopy()
+				// Optimistic-lock the patch so concurrent writes from sandbox-manager
+				// surface as 409 instead of silently winning a last-writer race.
+				patch := client.MergeFromWithOptions(box, client.MergeFromWithOptimisticLock{})
+				modified.Spec.Paused = true
+				if box.Spec.ShutdownTime != nil && box.Spec.ShutdownTime.After(now.Time) {
+					modified.Spec.PauseTime = box.Spec.ShutdownTime.DeepCopy()
+				} else {
+					farFuture := metav1.NewTime(now.AddDate(1000, 0, 0))
+					modified.Spec.PauseTime = &farFuture
```
Comment (furykerry) [inline pkg/controller/sandbox/sandbox_controller.go]: why set pauseTime to 1000 days ahead ? just clear pauseTime?

File: cmd/sandbox-manager/main.go [hunk 3]
```diff
@@ -80,6 +98,10 @@ func main() {
 	pflag.BoolVar(&e2bEnableAuth, "e2b-enable-auth", true, "Enable E2B authentication")
 	pflag.StringVar(&domain, "e2b-domain", "localhost", "E2B domain")
 	pflag.IntVar(&e2bMaxTimeout, "e2b-max-timeout", models.DefaultMaxTimeout, "E2B maximum timeout in seconds")
+	pflag.IntVar(&e2bMinResumeTimeout, "e2b-min-resume-timeout", models.DefaultMinResumeTimeoutSeconds,
+		"Minimum effective timeout (seconds) applied to Connect / Resume requests that trigger Resume on a Paused sandbox. "+
```
Comment (furykerry) [inline cmd/sandbox-manager/main.go]: consider rephrase the argument description to make the sentence more readable

File: pkg/servers/e2b/pause_resume.go [hunk 4]
```diff
@@ -108,47 +129,54 @@ func (sc *Controller) ResumeSandbox(r *http.Request) (web.ApiResponse[struct{}],
...
 		Code: http.StatusNoContent,
 	}, nil
 }
 
+// applyResumeFloor bumps `requested` up to sc.minResumeTimeout when the
+// request will trigger Resume on a timed Paused sandbox. The floor prevents
+// the placeholder PauseTime from expiring mid-Resume and triggering an
+// auto-pause race. Never-timeout sandboxes have no PauseTime so the floor
+// is skipped (and would only generate misleading log noise).
+func (sc *Controller) applyResumeFloor(log klog.Logger, requested int, paused, hasDeadline bool) int {
```
Comment (furykerry) [inline pkg/servers/e2b/pause_resume.go]: consider rename applyResumeFloor to getEffectivePauseTimeSeconds, applyResumeFloor is too hard to reason about
